File: ios/Sources/Scene/Confetti.swift
import RealityKit
import simd

/// A goal's confetti: cards thrown up from the net, tumbling down on real time. Presentation only —
/// its randomness is its own generator, never a match stream (§4.3). The twin of Confetti.kt.
@MainActor
final class Confetti {
    typealias C = Presentation.Celebration
    let root = Entity()
    private var pieces: [(e: ModelEntity, v: SIMD3<Float>, spin: SIMD3<Float>)] = []
    private var age = Double.infinity
    private var rng = SplitMixLite(seed: 0x5EED)

    init(materials: Materials, look: WorldLook) throws {
        let mesh = try Figures.card().resource()
        let colours = try C.colours.map { try materials.actor($0, look: look) }
        for i in 0..<C.pieces {
            let e = ModelEntity(mesh: mesh, materials: [colours[i % colours.count]])
            e.components.set(DynamicLightShadowComponent(castsShadow: false))
            e.isEnabled = false
            root.addChild(e)
            pieces.append((e, .zero, .zero))
        }
    }

    /// Throws the confetti up from the net on goal line `goalZ`.
    func burst(goalZ: Float) {
        age = 0
        let size = Float(C.size)
        for i in pieces.indices {
            let a = rng.next() * 2 * .pi
            let up = 0.55 + 0.45 * rng.next()
            let speed = Float(C.speed) * (0.6 + 0.4 * rng.next())
            let out: Float = goalZ >= 0 ? -1 : 1
            pieces[i].v = SIMD3(cos(a) * (1 - up), up, sin(a) * (1 - up) * 0.6 + out * 0.25) * speed
            pieces[i].spin = SIMD3(rng.next() - 0.5, rng.next() - 0.5, rng.next() - 0.5) * 14
            let e = pieces[i].e
            e.position = SIMD3((rng.next() - 0.5) * 6, 1.2, goalZ + (rng.next() - 0.5) * 1.5)
            e.scale = SIMD3(repeating: size)
            e.isEnabled = true
        }
    }

    /// Real time: fall, flutter, shrink away at the end.
    func advance(_ dt: Double) {
        guard age < C.seconds else { return }
        age += dt
        let t = Float(dt), g = Float(C.gravity)
        let fade = Float(max(0, min(1, (C.seconds - age) / 0.6)))
        for i in pieces.indices {
            var p = pieces[i]
            p.v.y -= g * t
            p.v *= exp(-1.4 * t)
            p.e.position += p.v * t
            if p.e.position.y < 0.05 { p.e.position.y = 0.05; p.v = .zero }
            p.e.orientation = simd_quatf(angle: simd_length(p.spin) * t, axis: simd_normalize(p.spin)) * p.e.orientation
            p.e.scale = SIMD3(repeating: Float(C.size) * fade)
            p.e.isEnabled = age < C.seconds
            pieces[i] = p
        }
    }
}

/// A small presentation-only generator (never on the simulation path, §4.3).
struct SplitMixLite {
    private var state: UInt64
    init(seed: UInt64) { state = seed }
    mutating func next() -> Float {
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return Float((z ^ (z >> 31)) >> 40) / Float(1 << 24)
    }
}

File: android/app/src/main/kotlin/in/nann/smashhockey/scene/Confetti.kt
package `in`.nann.smashhockey.scene

import com.google.android.filament.Engine
import com.google.android.filament.Scene
import `in`.nann.smashhockey.engine.Geometry
import `in`.nann.smashhockey.engine.Materials
import `in`.nann.smashhockey.engine.Node
import `in`.nann.smashhockey.generated.Presentation
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.sin

/**
 * A goal's confetti: cards thrown up from the net, tumbling down on real time — the twin of iOS's
 * Confetti.swift. Presentation only: its randomness is its own generator, never a match stream (§4.3).
 */
class Confetti(private val engine: Engine, private val scene: Scene, parent: Int, materials: Materials) {
    private val c = Presentation.Celebration
    private val geometry = Geometry(engine, Figures.card())
    private class Piece(val node: Node) { var vx = 0f; var vy = 0f; var vz = 0f; var sx = 0f; var sy = 0f; var sz = 0f }
    private val pieces: List<Piece>
    private var age = Double.POSITIVE_INFINITY
    private var state = 0x5EEDL
    private var visible = false

    init {
        val colours = c.colours.map { materials.actor(it) }
        pieces = (0 until c.pieces).map { i ->
            val e = geometry.renderable(listOf(colours[i % colours.size]), shadows = false)
            Piece(Node(engine, parent, existing = e))
        }
    }

    /** A presentation-only SplitMix64 draw in [0, 1). */
    private fun next(): Float {
        state += -0x61c8864680b583ebL
        var z = state
        z = (z xor (z ushr 30)) * -0x40a7b892e31b1a47L
        z = (z xor (z ushr 27)) * -0x6b2fb644ecceee15L
        return ((z xor (z ushr 31)) ushr 40).toFloat() / (1 shl 24).toFloat()
    }

    fun burst(goalZ: Float) {
        age = 0.0
        if (!visible) { pieces.forEach { scene.addEntity(it.node.entity) }; visible = true }
        for (p in pieces) {
            val a = next() * 2f * PI.toFloat()
            val up = 0.55f + 0.45f * next()
            val speed = c.speed.toFloat() * (0.6f + 0.4f * next())
            val out = if (goalZ >= 0) -1f else 1f
            p.vx = cos(a) * (1 - up) * speed; p.vy = up * speed; p.vz = (sin(a) * (1 - up) * 0.6f + out * 0.25f) * speed
            p.sx = (next() - 0.5f) * 14; p.sy = (next() - 0.5f) * 14; p.sz = (next() - 0.5f) * 14
            p.node.x = (next() - 0.5f) * 6; p.node.y = 1.2f; p.node.z = goalZ + (next() - 0.5f) * 1.5f
            p.node.scale(c.size.toFloat()); p.node.apply()
        }
    }

    fun advance(dt: Double) {
        if (age >= c.seconds) return
        age += dt
        if (age >= c.seconds) { pieces.forEach { scene.removeEntity(it.node.entity) }; visible = false; return }
        val t = dt.toFloat()
        val fade = ((c.seconds - age) / 0.6).coerceIn(0.0, 1.0).toFloat()
        val drag = exp(-1.4f * t)
        for (p in pieces) {
            p.vy -= c.gravity.toFloat() * t
            p.vx *= drag; p.vy *= drag; p.vz *= drag
            val n = p.node
            n.x += p.vx * t; n.y += p.vy * t; n.z += p.vz * t
            if (n.y < 0.05f) { n.y = 0.05f; p.vx = 0f; p.vy = 0f; p.vz = 0f }
            n.yaw += p.sy * t; n.pitch += p.sx * t; n.roll += p.sz * t
            n.scale(c.size.toFloat() * fade)
            n.apply()
        }
    }

    fun destroy() {
        if (visible) pieces.forEach { scene.removeEntity(it.node.entity) }
        geometry.destroy()
    }
}
